### PR TITLE
Two player support

### PR DESCRIPTION
**Pull request by kavanparam**
Alternating player selection is now working, and game win can be determined between two players!

- removed `GameBoardBlock`'s `block` state & extracted logic to App level passing down function as prop (`blockClick()`)
- UI:
       * font changes for buttons
       * added shading to GameBoard for depth/separation in the UI
       * removed unnecessary heading  
       * added different colours for player selection (selected blocks & hover state for block to be selected).
- added `isPlayerOne` state to manage current player state that’s toggled inside `blockClick()`
- refactored `isSelected` property of `GameBoardBlock` to enum `selectedBy`
- refactored all setStates to make sure no state was modified directly — used previousState and creating copies of state in setState calls
- refactored `calculateWinner()` to accommodate checking game win between two players

image

### CODE REVIEW

File: components/math/multiplicationConnect/GameBoard.tsx [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 import GameBoardBlock from "./GameBoardBlock";
 
 interface GameBoardProps {
   grid: GameBoardBlock[];
+  blockClick(block: GameBoardBlock): void;
+  isPlayerOne: boolean;
 }
 
-const GameBoard: FC<GameBoardProps> = ({ grid }) => {
+const GameBoard: FC<GameBoardProps> = ({ grid, blockClick, isPlayerOne }) => {
```
Comment (vithushan19) [inline components/math/multiplicationConnect/GameBoard.tsx:11]: Nice, the grid and isPlayerOne is something you will later refactor to a redux state.

This is great for now. You're currently doing what we call prop drilling. It's a valid technique, but can get messy with more complex components.

That's why we use redux to simplify this state management.

File: components/math/multiplicationConnect/GameBoardBlock.tsx [hunk 2]
```diff
@@ -1,66 +1,58 @@
...
-      Options to determine winner:
-        3. add an isPlayerOne variable to GameBoardBlock & evaluate accordingly in calculateWinner() — to make sure 4 blocks in a row are pressed by the same player
-  */
-  const [isPlayerOne, setIsPlayerOne] = useState(true);
+const GameBoardBlock: FC<GameBoardBlockProps> = ({
+  blockData,
+  blockClick,
+  isPlayerOne,
+}) => {
+  // console.log("Block", blockData);
```
Comment (vithushan19) [inline components/math/multiplicationConnect/GameBoardBlock.tsx:27]: Remove commented code

File: components/math/multiplicationConnect/GameBoardBlock.tsx [hunk 3]
```diff
@@ -1,66 +1,58 @@
...
+          ${
+            // Selected block colours
+            blockData.selectedBy === selectedBy.PlayerOne
+              ? "bg-[#F20000]/80"
+              : blockData.selectedBy === selectedBy.PlayerTwo
+              ? "bg-[#FFDB00]/90"
+              : ""
+          }
+          ${
+            // Unselected hover animations
```
Comment (vithushan19) [inline components/math/multiplicationConnect/GameBoardBlock.tsx:45]: Great comments

File: pages/studentPortal/labs/multiplication-connect/Index.tsx [hunk 4]
```diff
@@ -1,12 +1,21 @@
...
 import React, { FC, useState, useEffect } from "react";
 import DiceSection from "../../../../components/math/multiplicationConnect/DiceSection";
 import GameBoard from "../../../../components/math/multiplicationConnect/GameBoard";
 import GameBoardBlock from "../../../../components/math/multiplicationConnect/GameBoardBlock";
 import PlayerSection from "../../../../components/math/multiplicationConnect/PlayerSection";
 import { getRandomItemFromArray } from "../../../api/random";
 
-export const calculateWinner = (grid: GameBoardBlock[]) => {
-  // Algorithm ran on block click to determine win
+enum selectedBy {
```
Comment (vithushan19) [inline pages/studentPortal/labs/multiplication-connect/Index.tsx:9]: Ah I see. You actually have a type called selectedBy.

I would capitalize the s here

File: components/math/multiplicationConnect/GameBoardBlock.tsx [hunk 5]
```diff
@@ -1,66 +1,58 @@
...
+  Unselected = "UNSELECTED",
+  PlayerOne = "PLAYERONE",
+  PlayerTwo = "PLAYERTWO",
+}
+
 interface GameBoardBlock {
   id: number;
   gridNumber: number;
-  isSelected: boolean;
+  selectedBy: selectedBy;
```
Comment (vithushan19) [inline components/math/multiplicationConnect/GameBoardBlock.tsx:13]: I would capitalize the s for the type

File: pages/studentPortal/labs/multiplication-connect/Index.tsx [hunk 6]
```diff
@@ -1,12 +1,21 @@
...
+  Unselected = "UNSELECTED",
+  PlayerOne = "PLAYERONE",
+  PlayerTwo = "PLAYERTWO",
+}
+
+export const calculateWinner = (
```
Comment (vithushan19) [inline pages/studentPortal/labs/multiplication-connect/Index.tsx:15]: I would write a unit test for this function

File: pages/studentPortal/labs/multiplication-connect/Index.tsx [hunk 7]
```diff
@@ -16,106 +25,124 @@ export const calculateWinner = (grid: GameBoardBlock[]) => {
...
-    }
-
-    return newGrid;
-  };
-
-  // useEffect not triggering re-render of GameBoard even though grid State is being updated
-  // grid state may need to included in dependency array but this results in infinite loop of renders
   useEffect(() => {
     console.log(`newGame: ${newGame}`);
     setGrid(createGrid);
```
Comment (vithushan19) [inline pages/studentPortal/labs/multiplication-connect/Index.tsx:100]: Did you figure out why this effect wasn't working?

**Review by vithushan19 — COMMENTED** [review]
Very nice!

File: pages/studentPortal/labs/multiplication-connect/Index.tsx (hunk 7 above)
Comment (kavanparam) [reply pages/studentPortal/labs/multiplication-connect/Index.tsx:100]: Yeah I worked thorough it with Abhi and Jason and managed to figure it out. Jason got it working by adding another useEffect in GameBoardBlock to update the blocks when the data passed to them as a prop changes (on game reset). Bc the grid was updating but the individual block states in the child component wasn't.

But then he mentioned the implementation was kind of messy and he would skip the `block` state altogether and just manage the `grid` state instead. So I just reworked the code bc that made more sense.

**Pull request merged**